ionoscloud/resource_dbaas_inmemorydb_replicaset.go

```diff
@@ -0,0 +1,291 @@
+										Type:     schema.TypeString,
+										Required: true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			"maintenance_window": {
+				Type:        schema.TypeList,
+				Description: "A weekly 4 hour-long window, during which maintenance might occur.",
+				Optional:    true,
+				Computed:    true,
+				MaxItems:    1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"time": {
+							Type:        schema.TypeString,
+							Description: "Start of the maintenance window in UTC time.",
+							Required:    true,
+						},
+						"day_of_the_week": {
+							Type:        schema.TypeString,
+							Description: "The name of the week day.",
+							Required:    true,
+						},
+					},
+				},
+			},
+			"initial_snapshot_id": {
+				Type:        schema.TypeString,
+				Description: "The ID of a snapshot to restore the replica set from. If set, the replica set will be created from the snapshot.",
+				Optional:    true,
+			},
+			"dns_name": {
+				Type:        schema.TypeString,
+				Description: "The DNS name pointing to your replica set. Will be used to connect to the active/standalone instance.",
+				Computed:    true,
+			},
+		},
+		Timeouts: &resourceDefaultTimeouts,
+	}
+}
+
+func replicaSetCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(services.SdkBundle).InMemoryDBClient
+
+	replicaSet := inmemorydb.GetReplicaSetDataCreate(d)
+	response, _, err := client.CreateReplicaSet(ctx, replicaSet, d.Get("location").(string))
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("an error occurred while creating an InMemoryDB replica set: %w", err))
+	}
+	replicaSetID := *response.Id
+	d.SetId(replicaSetID)
+	err = utils.WaitForResourceToBeReady(ctx, d, client.IsReplicaSetReady)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error occurred while checking the status for InMemoryDB replica set with ID: %v, error: %w", replicaSetID, err))
+	}
+	// Call the read function to save the DNS name in the state (DNS name is not present in the creation response).
+	return replicaSetRead(ctx, d, meta)
+}
+
+func replicaSetDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(services.SdkBundle).InMemoryDBClient
+	replicaSetID := d.Id()
+	apiResponse, err := client.DeleteReplicaSet(ctx, replicaSetID, d.Get("location").(string))
+	if err != nil {
+		if apiResponse.HttpNotFound() {
+			d.SetId("")
+			return nil
+		}
+		return diag.FromErr(fmt.Errorf("error while deleting InMemoryDB replica set with ID: %v, error: %w", replicaSetID, err))
+	}
+	err = utils.WaitForResourceToBeDeleted(ctx, d, client.IsReplicaSetDeleted)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("deletion check failed for InMemoryDB replica set with ID: %v, error: %w", replicaSetID, err))
+	}
+
+	// wait for the lan to be freed after the deletion of the replica set
+	time.Sleep(constant.SleepInterval * 10)
```

Comment:
is there any way of using a "WaitUntil..." type function instead of an arbitrary sleep ?

Reply:
@cbeti-ionos Yes, but it would mean to involve two clients, one for the InMemoryDB and one for CloudAPI since I will need to fetch the LAN and to check that the LAN is actually deleted. I will try this once the API problem related to the LAN deletion is fixed.